File: app/src/main/java/net/ducksmanager/whattheduck/CountryAdapter.java
package net.ducksmanager.whattheduck;

import android.app.Activity;
import android.content.Intent;
import android.view.View;

import net.ducksmanager.persistence.models.composite.InducksCountryNameWithPossession;

import java.util.List;

import androidx.recyclerview.widget.RecyclerView;

public class CountryAdapter extends ItemAdapter<InducksCountryNameWithPossession> {

    CountryAdapter(ItemList itemList, List<InducksCountryNameWithPossession> items) {
        super(itemList, R.layout.row, items);
    }

    @Override
    protected ViewHolder getViewHolder(View v) {
        return new ViewHolder(v);
    }

    @Override
    protected boolean isPossessed(InducksCountryNameWithPossession c) {
        return c.isPossessed();
    }

    @Override
    protected View.OnClickListener getOnClickListener() {
        return view -> {
            int position = ((RecyclerView)view.getParent()).getChildLayoutPosition(view);
            InducksCountryNameWithPossession selectedCountry = CountryAdapter.this.getItem(position);
            WhatTheDuck.selectedCountry = selectedCountry.country.getCountryCode();

            Intent i = new Intent(originActivity, PublicationList.class);
            originActivity.startActivity(i);
        };
    }

    class ViewHolder extends ItemAdapter.ViewHolder {
        ViewHolder(View v) {
            super(v);
        }
    }

    @Override
    protected Integer getPrefixImageResource(InducksCountryNameWithPossession c, Activity a) {
        String uri = "@drawable/flags_" + c.country.getCountryCode();
        int imageResource = a.getResources().getIdentifier(uri, null, a.getPackageName());

        if (imageResource == 0) {
            imageResource = R.drawable.flags_unknown;
        }
        return imageResource;
    }

    @Override
    protected Integer getSuffixImageResource(InducksCountryNameWithPossession c) {
        return null;
    }

    @Override
    protected String getSuffixText(InducksCountryNameWithPossession c) {
        return null;
    }

    @Override
    protected String getText(InducksCountryNameWithPossession c) {
        return c.country.getCountryName();
    }

    @Override
    protected String getIdentifier(InducksCountryNameWithPossession c) {
        return c.country.getCountryCode();
    }

    @Override
    protected String getComparatorText(InducksCountryNameWithPossession c) {
        return getText(c);
    }
}

File: app/src/main/java/net/ducksmanager/whattheduck/CountryAdapter.kt
package net.ducksmanager.whattheduck

import android.app.Activity
import android.content.Intent
import android.support.v7.widget.RecyclerView
import android.view.View

import java.util.ArrayList

class CountryAdapter internal constructor(itemList: ItemList<*>, items: ArrayList<Country>) : ItemAdapter<CountryAdapter.Country>(itemList, R.layout.row, items) {

    protected override val onClickListener: View.OnClickListener
        get() = { view ->
            val position = (view.getParent() as RecyclerView).getChildLayoutPosition(view)
            val selectedCountry = this@CountryAdapter.getItem(position)
            WhatTheDuck.selectedCountry = selectedCountry.shortName

            val i = Intent(originActivity, PublicationList::class.java)
            originActivity.startActivity(i)
        }

    internal class Country(val shortName: String, val fullName: String)


    override fun getViewHolder(v: View): ViewHolder {
        return ViewHolder(v)
    }

    protected inner class ViewHolder internal constructor(v: View) : ItemAdapter.ViewHolder(v)

    override fun isHighlighted(i: Country): Boolean {
        return WhatTheDuck.userCollection.hasCountry(i.shortName)
    }

    override fun getPrefixImageResource(i: Country, a: Activity): Int? {
        val uri = "@drawable/flags_" + i.shortName
        var imageResource = a.resources.getIdentifier(uri, null, a.packageName)

        if (imageResource == 0) {
            imageResource = R.drawable.flags_unknown
        }
        return imageResource
    }

    override fun getSuffixImageResource(i: Country): Int? {
        return null
    }

    override fun getSuffixText(i: Country): String? {
        return null
    }

    override fun getText(i: Country): String {
        return i.fullName
    }

    override fun getIdentifier(i: Country): String {
        return i.shortName
    }

    override fun getComparatorText(i: Country): String? {
        return getText(i)
    }
}
